### ISSUE TITLE
The domain classifier loss is not decreasing.

### ISSUE THREAD

**Issue by zhongpeixiang** [fungtion/DANN_py3#12]
<img width="1094" alt="image" src="https://user-images.githubusercontent.com/11826803/224246135-f5b2d9a0-986d-4aec-9a23-93e153028e02.png">

As shown in the image above, the domain classifier loss is almost constant throughout the training process. I use a ViT as the feature extractor, a linear layer as label classifier, and a two-layer MLP as the domain classifier. 

What are the possible causes? and what are the typical loss curves for domain classifier?

Thanks

**Comment by cs-mshah** [comment]
Even I am facing this issue. The DANN loss converges quite fast. Here are the plots of this repository without any changes to the code: [wandb-DANN](https://wandb.ai/manan-shah/DANN?workspace=user-manan-shah). In my other case I tried using a resnet18 and the domain classifier similar to the classifier part of resnet. There also the loss almost instantly stabilised to around 0.7. Is there any good repository where we can clearly understand how DANN works and a CLEAR practical working example of stable training?

**Comment by taotaowang97479** [comment]
> Even I am facing this issue. The DANN loss converges quite fast. Here are the plots of this repository without any changes to the code: [wandb-DANN](https://wandb.ai/manan-shah/DANN?workspace=user-manan-shah). In my other case I tried using a resnet18 and the domain classifier similar to the classifier part of resnet. There also the loss almost instantly stabilised to around 0.7. Is there any good repository where we can clearly understand how DANN works and a CLEAR practical working example of stable training?

I had the same problem. I used the code and data set given by the author, and there was no problem in the training. With the progress of training, the loss of the domain classifier stabilized at 0.65-0.67, and the training loss was shown in the figure. 
![train_loss](https://github.com/fungtion/DANN_py3/assets/138740461/c707225a-85ad-4c66-8024-92b599df089e)
However, with my own network, the same training method, the domain classifier loss is stable at 0.69 from the beginning, which seems to indicate that the domain classifier is not learning, it will randomly classify the source domain or target domain samples with 50% probability.
Can anyone figure out how to fix it?

**Comment by cs-mshah** [comment]
I had shifted to a more robust codebase: https://github.com/thuml/Transfer-Learning-Library

**Comment by taotaowang97479** [comment]
> I had shifted to a more robust codebase: https://github.com/thuml/Transfer-Learning-Library

So is it the code itself? I went to this codabase and looked at the same DANN code and I didn't think there was a big difference in the way it was written

**Comment by JialingRichard** [comment]
> > Even I am facing this issue. The DANN loss converges quite fast. Here are the plots of this repository without any changes to the code: [wandb-DANN](https://wandb.ai/manan-shah/DANN?workspace=user-manan-shah). In my other case I tried using a resnet18 and the domain classifier similar to the classifier part of resnet. There also the loss almost instantly stabilised to around 0.7. Is there any good repository where we can clearly understand how DANN works and a CLEAR practical working example of stable training?
> 
> I had the same problem. I used the code and data set given by the author, and there was no problem in the training. With the progress of training, the loss of the domain classifier stabilized at 0.65-0.67, and the training loss was shown in the figure. ![train_loss](https://private-user-images.githubusercontent.com/138740461/329452658-c707225a-85ad-4c66-8024-92b599df089e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2OTI4OTksIm5iZiI6MTczMzY5MjU5OSwicGF0aCI6Ii8xMzg3NDA0NjEvMzI5NDUyNjU4LWM3MDcyMjVhLTg1YWQtNGM2Ni04MDI0LTkyYjU5OWRmMDg5ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwOFQyMTE2MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NGQ1ZjgyZDEwMzUxZWVmMjkxNTc1MDI4NWJmOWU0NDI4ZTMyOTBjOWFmOTkyZjAzM2U2OGFjMTU4OGUyY2ZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.HgCSRJvNaWoF9lzRSSK0W0-5hVv9wK9hDMq4fWQ85IY) However, with my own network, the same training method, the domain classifier loss is stable at 0.69 from the beginning, which seems to indicate that the domain classifier is not learning, it will randomly classify the source domain or target domain samples with 50% probability. Can anyone figure out how to fix it?

Does this means the domain classifier even didn't learn anything? Did you solve this problem， like other resp's code can be better?

**Comment by taotaowang97479** [comment]
> > > Even I am facing this issue. The DANN loss converges quite fast. Here are the plots of this repository without any changes to the code: [wandb-DANN](https://wandb.ai/manan-shah/DANN?workspace=user-manan-shah). In my other case I tried using a resnet18 and the domain classifier similar to the classifier part of resnet. There also the loss almost instantly stabilised to around 0.7. Is there any good repository where we can clearly understand how DANN works and a CLEAR practical working example of stable training?
> > 
> > 
> > I had the same problem. I used the code and data set given by the author, and there was no problem in the training. With the progress of training, the loss of the domain classifier stabilized at 0.65-0.67, and the training loss was shown in the figure. ![train_loss](https://private-user-images.githubusercontent.com/138740461/329452658-c707225a-85ad-4c66-8024-92b599df089e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2OTI4OTksIm5iZiI6MTczMzY5MjU5OSwicGF0aCI6Ii8xMzg3NDA0NjEvMzI5NDUyNjU4LWM3MDcyMjVhLTg1YWQtNGM2Ni04MDI0LTkyYjU5OWRmMDg5ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwOFQyMTE2MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NGQ1ZjgyZDEwMzUxZWVmMjkxNTc1MDI4NWJmOWU0NDI4ZTMyOTBjOWFmOTkyZjAzM2U2OGFjMTU4OGUyY2ZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.HgCSRJvNaWoF9lzRSSK0W0-5hVv9wK9hDMq4fWQ85IY) However, with my own network, the same training method, the domain classifier loss is stable at 0.69 from the beginning, which seems to indicate that the domain classifier is not learning, it will randomly classify the source domain or target domain samples with 50% probability. Can anyone figure out how to fix it?
> 
> Does this means the domain classifier even didn't learn anything? Did you solve this problem， like other resp's code can be better?

Yes, the domain classifier has not learned anything. I have already abandoned my project, as it is not an issue with the code itself. Based on my experience, hyperparameter tuning does not have much effect (I have tried random search for hyperparameters). I believe the key to successful training is whether the dataset aligns with the network architecture: if the data is too challenging, the network is likely unable to learn domain shift; if the network architecture is overly complex, adversarial training is easily unbalanced. Both of these factors can lead to the domain classifier performing random classification.

**Comment by JialingRichard** [comment]
> > > > Even I am facing this issue. The DANN loss converges quite fast. Here are the plots of this repository without any changes to the code: [wandb-DANN](https://wandb.ai/manan-shah/DANN?workspace=user-manan-shah). In my other case I tried using a resnet18 and the domain classifier similar to the classifier part of resnet. There also the loss almost instantly stabilised to around 0.7. Is there any good repository where we can clearly understand how DANN works and a CLEAR practical working example of stable training?
> > > 
> > > 
> > > I had the same problem. I used the code and data set given by the author, and there was no problem in the training. With the progress of training, the loss of the domain classifier stabilized at 0.65-0.67, and the training loss was shown in the figure. ![train_loss](https://private-user-images.githubusercontent.com/138740461/329452658-c707225a-85ad-4c66-8024-92b599df089e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2OTI4OTksIm5iZiI6MTczMzY5MjU5OSwicGF0aCI6Ii8xMzg3NDA0NjEvMzI5NDUyNjU4LWM3MDcyMjVhLTg1YWQtNGM2Ni04MDI0LTkyYjU5OWRmMDg5ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwOFQyMTE2MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NGQ1ZjgyZDEwMzUxZWVmMjkxNTc1MDI4NWJmOWU0NDI4ZTMyOTBjOWFmOTkyZjAzM2U2OGFjMTU4OGUyY2ZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.HgCSRJvNaWoF9lzRSSK0W0-5hVv9wK9hDMq4fWQ85IY) However, with my own network, the same training method, the domain classifier loss is stable at 0.69 from the beginning, which seems to indicate that the domain classifier is not learning, it will randomly classify the source domain or target domain samples with 50% probability. Can anyone figure out how to fix it?
> > 
> > 
> > Does this means the domain classifier even didn't learn anything? Did you solve this problem， like other resp's code can be better?
> 
> Yes, the domain classifier has not learned anything. I have already abandoned my project, as it is not an issue with the code itself. Based on my experience, hyperparameter tuning does not have much effect (I have tried random search for hyperparameters). I believe the key to successful training is whether the dataset aligns with the network architecture: if the data is too challenging, the network is likely unable to learn domain shift; if the network architecture is overly complex, adversarial training is easily unbalanced. Both of these factors can lead to the domain classifier performing random classification.

Hi, I just did a training use this code this week. I found maybe we can set alpha(the weight of GRL layer) is very small at first,  to make sure the classifier is strong enough, then we start scale the alpha to make sure the feature extractor layer start 'cheating'. As a result, the accuracy of source and domain classifier accuracy can be very high at first then slow down after that finally they both can be very close to 0.5 accuracy,  just like this picture below (red and purple line for 2 domain classifier accuracy). So that We can have a better performance, although this changes of alpha are completed by a math function in this resp code. Hope this can be useful for you to train DANN maybe.
![图片](https://github.com/user-attachments/assets/061109ff-def5-49f8-a057-139d9ca3d1ea)

**Comment by taotaowang97479** [comment]
> > > > > Even I am facing this issue. The DANN loss converges quite fast. Here are the plots of this repository without any changes to the code: [wandb-DANN](https://wandb.ai/manan-shah/DANN?workspace=user-manan-shah). In my other case I tried using a resnet18 and the domain classifier similar to the classifier part of resnet. There also the loss almost instantly stabilised to around 0.7. Is there any good repository where we can clearly understand how DANN works and a CLEAR practical working example of stable training?
> > > > 
> > > > 
> > > > I had the same problem. I used the code and data set given by the author, and there was no problem in the training. With the progress of training, the loss of the domain classifier stabilized at 0.65-0.67, and the training loss was shown in the figure. ![train_loss](https://private-user-images.githubusercontent.com/138740461/329452658-c707225a-85ad-4c66-8024-92b599df089e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2OTI4OTksIm5iZiI6MTczMzY5MjU5OSwicGF0aCI6Ii8xMzg3NDA0NjEvMzI5NDUyNjU4LWM3MDcyMjVhLTg1YWQtNGM2Ni04MDI0LTkyYjU5OWRmMDg5ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwOFQyMTE2MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NGQ1ZjgyZDEwMzUxZWVmMjkxNTc1MDI4NWJmOWU0NDI4ZTMyOTBjOWFmOTkyZjAzM2U2OGFjMTU4OGUyY2ZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.HgCSRJvNaWoF9lzRSSK0W0-5hVv9wK9hDMq4fWQ85IY) However, with my own network, the same training method, the domain classifier loss is stable at 0.69 from the beginning, which seems to indicate that the domain classifier is not learning, it will randomly classify the source domain or target domain samples with 50% probability. Can anyone figure out how to fix it?
> > > 
> > > 
> > > Does this means the domain classifier even didn't learn anything? Did you solve this problem， like other resp's code can be better?
> > 
> > 
> > Yes, the domain classifier has not learned anything. I have already abandoned my project, as it is not an issue with the code itself. Based on my experience, hyperparameter tuning does not have much effect (I have tried random search for hyperparameters). I believe the key to successful training is whether the dataset aligns with the network architecture: if the data is too challenging, the network is likely unable to learn domain shift; if the network architecture is overly complex, adversarial training is easily unbalanced. Both of these factors can lead to the domain classifier performing random classification.
> 
> Hi, I just did a training use this code this week. I found maybe we can set alpha(the weight of GRL layer) is very small at first, to make sure the classifier is strong enough, then we start scale the alpha to make sure the feature extractor layer start 'cheating'. As a result, the accuracy of source and domain classifier accuracy can be very high at first then slow down after that finally they both can be very close to 0.5 accuracy, just like this picture below (red and purple line for 2 domain classifier accuracy). So that We can have a better performance, although this changes of alpha are completed by a math function in this resp code. Hope this can be useful for you to train DANN maybe. ![图片](https://private-user-images.githubusercontent.com/81303390/395654749-061109ff-def5-49f8-a057-139d9ca3d1ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMTgzNDEsIm5iZiI6MTczNDMxODA0MSwicGF0aCI6Ii84MTMwMzM5MC8zOTU2NTQ3NDktMDYxMTA5ZmYtZGVmNS00OWY4LWEwNTctMTM5ZDljYTNkMWVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE2VDAzMDA0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0MWViZTY2NTQ5MTI0YzY4ZTI1OTAwMjUzNWNiYzhhNzhkNGEzZjg3NDUzNmZkZDBmZmZhN2MyOGI4YTcxNmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HbL6gMiouTUNnO3GyK50rKmzAjDH7p4OUHd4ggVfNi4)

Hi, your ACC and the loss in the image I shared are perfectly matched in this DANN project. In some of the applications of DANN in my field, some alphas are fixed throughout, while others, like in this DANN project, gradually increase as the number of training iterations increases. When I previously used random search for parameters, I always used a fixed alpha value. Now, I'm trying to dynamically adjust alpha. Thanks for your reminder.

**Comment by cakutoo** [comment]
I used the DANN architecture, where the domain classification loss was combined with another accuracy-related loss function through addition. A GRL (Gradient Reversal Layer) was applied to reverse the gradients at the input of the domain classifier. However, the domain classification loss kept decreasing and never showed any sign of increasing.

**Comment by jiaweiww** [comment]
Has anyone tried using binomial cross-entropy loss to replace the NLLLoss?
